internal/statesync/reactor.go

```diff
@@ -831,20 +835,41 @@ func (r *Reactor) handleMessage(ctx context.Context, chID p2p.ChannelID, envelop
 // encountered during message execution will result in a PeerError being sent on
 // the respective channel. When the reactor is stopped, we will catch the signal
 // and close the p2p Channel gracefully.
-func (r *Reactor) processCh(ctx context.Context, ch *p2p.Channel, chName string) {
-	iter := ch.Receive(ctx)
+func (r *Reactor) processChannels(ctx context.Context, chs ...*p2p.Channel) {
+	// make sure that the iterator get cleaned up correctly in the
+	// case of an error
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	chanTable := make(map[conn.ChannelID]*p2p.Channel, len(chs))
+	for idx := range chs {
+		ch := chs[idx]
+		chanTable[ch.ID] = ch
+	}
+
+	iter := p2p.MergedChannelIterator(ctx, chs...)
 	for iter.Next(ctx) {
 		envelope := iter.Envelope()
-		if err := r.handleMessage(ctx, ch.ID, envelope); err != nil {
-			r.logger.Error("failed to process message",
-				"err", err,
-				"channel", chName,
-				"ch_id", ch.ID,
-				"envelope", envelope)
-			if serr := ch.SendError(ctx, p2p.PeerError{
-				NodeID: envelope.From,
-				Err:    err,
-			}); serr != nil {
+		if err := r.handleMessage(ctx, envelope); err != nil {
+			if ch, ok := chanTable[envelope.ChannelID]; ok {
+				r.logger.Error("failed to process message",
+					"err", err,
+					"channel", ch.String(),
+					"ch_id", envelope.ChannelID,
+					"envelope", envelope)
+				if serr := ch.SendError(ctx, p2p.PeerError{
+					NodeID: envelope.From,
+					Err:    err,
+				}); serr != nil {
+					return
+				}
+			} else {
+				r.logger.Error("received impossible message",
+					"envelope_from", envelope.From,
+					"envelope_ch", envelope.ChannelID,
+					"num_chs", len(chanTable),
+					"err", err,
+				)
 				return
 			}
```

Comment:
Some of this is pre-existing, but I think we could comfortably [indent error flow](https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow) here.
```suggestion
			ch, ok := chanTable[envelope.ChannelID]
			if !ok {
				r.logger.Error("received impossible message",
					"envelope_from", envelope.From,
					"envelope_ch", envelope.ChannelID,
					"num_chs", len(chanTable),
					"err", err,
				)
				return
			}
			r.logger.Error("failed to process message",
				"err", err,
				"channel", ch.String(),
				"ch_id", envelope.ChannelID,
				"envelope", envelope)
			if serr := ch.SendError(ctx, p2p.PeerError{
				NodeID: envelope.From,
				Err:    err,
			}); serr != nil {
				return
			}
```